### PR TITLE
feat(rnd,infra) Set up terraform

### PR DESCRIPTION
**Pull request by aarushik93**
### Background
First base set up for k8s

### Changes 🏗️

Added terraform for k8s and networking.

will follow up with iams

### PR Quality Scorecard ✨

- [x] Have you used the PR description template? &ensp; `+2 pts`
- [x] Is your pull request atomic, focusing on a single change? &ensp; `+5 pts`
- [ ] Have you linked the GitHub issue(s) that this PR addresses? &ensp; `+5 pts`
- [x] Have you documented your changes clearly and comprehensively? &ensp; `+5 pts`
- [ ] Have you changed or added a feature? &ensp; `-4 pts`
  - [ ] Have you added/updated corresponding documentation? &ensp; `+4 pts`
  - [ ] Have you added/updated corresponding integration tests? &ensp; `+5 pts`
- [ ] Have you changed the behavior of AutoGPT? &ensp; `-5 pts`
  - [ ] Have you also run `agbenchmark` to verify that these changes do not regress performance? &ensp; `+10 pts`

___

### **PR Type**
Enhancement, Configuration changes

___

### **Description**
- Added a new GitHub Actions workflow for Terraform linting.
- Configured Terraform settings for Google Cloud, including provider and version requirements.
- Defined Terraform modules for networking and GKE cluster.
- Added resource definitions for VPC network, subnetwork, and GKE cluster.
- Configured variables and outputs for both networking and GKE cluster modules.
- Included Terraform lock file for managing provider versions.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Configuration changes</strong></td><td><details><summary>11 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>autogpt-infra-ci.yml</strong><dd><code>Add GitHub Actions workflow for Terraform linting</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

.github/workflows/autogpt-infra-ci.yml

<li>Added a new GitHub Actions workflow for Terraform linting.<br> <li> Configured the workflow to run on pushes and pull requests affecting <br>specific paths.<br> <li> Set up the job to use TFLint for linting Terraform code.<br>

</details>

  </td>
  <td><a href="https://github.com/Significant-Gravitas/AutoGPT/pull/7565/files#diff-9158fb95e3bb80355bd77494437e06bf76c75cb84c083122f656e886be15ae72">+33/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>.terraform.lock.hcl</strong><dd><code>Add Terraform lock file for provider version management</code>&nbsp; &nbsp; </dd></summary>
<hr>

rnd/infra/terraform/.terraform.lock.hcl

- Added Terraform lock file to manage provider versions.

</details>

  </td>
  <td><a href="https://github.com/Significant-Gravitas/AutoGPT/pull/7565/files#diff-5c15a8464e5fd297eccada462cc409fb1d9b839bf83077dd1efb783f18b375eb">+22/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>dev.tfvars</strong><dd><code>Add Terraform variables for development environment</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rnd/infra/terraform/environments/dev.tfvars

<li>Added Terraform variables for the development environment.<br> <li> Configured project ID, region, zone, network, subnet, cluster, node <br>pool, machine type, and disk size.<br>

</details>

  </td>
  <td><a href="https://github.com/Significant-Gravitas/AutoGPT/pull/7565/files#diff-afbe1103c303a8001bc8c27865476c0e3d97707e560b5b023a627798504f5e38">+11/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>main.tf</strong><dd><code>Configure main Terraform settings and modules</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rnd/infra/terraform/main.tf

<li>Defined required Terraform version and providers.<br> <li> Configured Google Cloud provider.<br> <li> Added modules for networking and GKE cluster.<br>

</details>

  </td>
  <td><a href="https://github.com/Significant-Gravitas/AutoGPT/pull/7565/files#diff-0ed6465c6e4b9f7ee3af0bc4cf5eb58d4c2cefa1ad156cc6ee969bfd82cf5820">+39/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>main.tf</strong><dd><code>Add GKE cluster resource definition</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rnd/infra/terraform/modules/gke_cluster/main.tf

<li>Added resource definition for Google Kubernetes Engine (GKE) cluster.<br> <li> Configured node pool settings based on variables.<br>

</details>

  </td>
  <td><a href="https://github.com/Significant-Gravitas/AutoGPT/pull/7565/files#diff-89c8bcdf781ef0ef53fdf84c0043ad0cc42ef05eef4e2e1960140c27ca749603">+21/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>outputs.tf</strong><dd><code>Add outputs for GKE cluster module</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rnd/infra/terraform/modules/gke_cluster/outputs.tf

<li>Defined outputs for GKE cluster name, endpoint, and node pool name.<br>

</details>

  </td>
  <td><a href="https://github.com/Significant-Gravitas/AutoGPT/pull/7565/files#diff-dccbe9a57de3d7910b01edf3dfed608095afb271bb46256b1b6f0ef6c1b7edcd">+14/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>variables.tf</strong><dd><code>Define variables for GKE cluster module</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rnd/infra/terraform/modules/gke_cluster/variables.tf

- Added variables for GKE cluster configuration.

</details>

  </td>
  <td><a href="https://github.com/Significant-Gravitas/AutoGPT/pull/7565/files#diff-c014fbb65275e449faeaf23d544a55ff787b90eb1633b1723e45d36b25f29019">+41/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>main.tf</strong><dd><code>Add VPC network and subnetwork resources</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rnd/infra/terraform/modules/networking/main.tf

- Added resource definitions for VPC network and subnetwork.

</details>

  </td>
  <td><a href="https://github.com/Significant-Gravitas/AutoGPT/pull/7565/files#diff-1abe628a79ad15289ee40c3ce0fb7406beda15e6dcc043212c5e33ce4bcf3bfd">+12/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>ouputs.tf</strong><dd><code>Add outputs for networking module</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rnd/infra/terraform/modules/networking/ouputs.tf

- Defined outputs for network and subnet details.

</details>

  </td>
  <td><a href="https://github.com/Significant-Gravitas/AutoGPT/pull/7565/files#diff-aa66ec103140624ad3271bbf8682c489fb193a6ca977ef3149c361ba7dbf415c">+19/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>variables.tf</strong><dd><code>Define variables for networking module</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rnd/infra/terraform/modules/networking/variables.tf

- Added variables for networking configuration.

</details>

  </td>
  <td><a href="https://github.com/Significant-Gravitas/AutoGPT/pull/7565/files#diff-f66cdcd16bea11d28230833373bfa8f22132d876771da96fcb37e34ac2c58a29">+21/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>variables.tf</strong><dd><code>Define global variables for Terraform configuration</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rnd/infra/terraform/variables.tf

<li>Added global variables for Terraform configuration.<br> <li> Configured project, region, zone, network, subnet, cluster, node pool, <br>machine type, disk size, and autopilot settings.<br>

</details>

  </td>
  <td><a href="https://github.com/Significant-Gravitas/AutoGPT/pull/7565/files#diff-b6aaf3634e43bd484993882fcc30b1c1ffc70b0e21bee8e0d129bfb52a444f80">+70/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: rnd/infra/terraform/main.tf [hunk 1]
```diff
@@ -0,0 +1,39 @@
+terraform {
+  required_version = ">= 1.9.0"
+  required_providers {
+    google = {
+      source  = "hashicorp/google"
+      version = "~> 4.0"
+    }
+  }
+}
```
Comment (Swiftyos) [inline rnd/infra/terraform/main.tf:13]: Can we store the tfstate in a gcs bucket?

```
terraform {
  required_version = ">= 1.9.0"
  required_providers {
    google = {
      source  = "hashicorp/google"
      version = "~> 4.0"
    }
  }

  backend "gcs" {
    bucket = "agpt-dev-terraform"
    prefix  = "terraform/state"
  }
}

File: .github/workflows/autogpt-infra-ci.yml [hunk 2]
```diff
@@ -0,0 +1,33 @@
+name: AutoGPT Builder CI
```
Comment (ntindle) [inline .github/workflows/autogpt-infra-ci.yml]:
```suggestion
name: AutoGPT Builder Infra
```

File: rnd/infra/terraform/.terraform.lock.hcl [hunk 3]
```diff
@@ -0,0 +1,22 @@
+# This file is maintained automatically by "terraform init".
```
Comment (Swiftyos) [inline rnd/infra/terraform/.terraform.lock.hcl:1]: Should we be committing this file?

File: rnd/infra/terraform/environments/dev.tfvars [hunk 4]
```diff
@@ -0,0 +1,11 @@
+project_id     = "agpt-dev"
```
Comment (Swiftyos) [inline rnd/infra/terraform/environments/dev.tfvars:1]: This is fine how it is, however my preference is to use white space to give a visual grouping of what variables are logically grouped together. Such as general, networking or cluster ones

**Review by Swiftyos — APPROVED** [review]

**Comment by Swiftyos** [comment]
Nice work!

File: rnd/infra/terraform/.terraform.lock.hcl (hunk 3 above)
Comment (aarushik93) [reply rnd/infra/terraform/.terraform.lock.hcl:1]: yup, we need these to maintain versions. probably one thing we could do is also lockdown platform specific - but I didn't cause ideally we'll do terraforming via CI and never on our own machines

File: rnd/infra/terraform/environments/dev.tfvars (hunk 4 above)
Comment (aarushik93) [reply rnd/infra/terraform/environments/dev.tfvars:1]: yeah fair call - as we end up with more modules, I was gonna re do how we do environments.. I think that will help the visual readability. but if you prefer, I can also just add the white space now lmk

**Comment by aarushik93** [comment]
@Swiftyos I'm gonna merge this PR to keep this stuff moving, I have a few more PRs coming - if you want the white spaces lmk and ill add in a second PR

**Pull request merged**